### DISCUSSION TITLE
Atoms "sometimes async" illegal in TypeScript

### DISCUSSION THREAD

**Discussion by fricze** [pmndrs/jotai#1589] [Q&A] [answered]
## Summary
Atom that starts as sync cannot be properly used as async in TypeScript. Based on [Async sometimes](https://jotai.org/docs/guides/async#async-sometimes) part in documentation any atom, even originally sync, should be able to receive `Promise` as the new value and act "async" then.

```
const request = async () => fetch('https://...').then((res) => res.json())
const baseAtom = atom(0)
const Component = () => {
  const [value, setValue] = useAtom(baseAtom)
  const handleClick = () => {
    setValue(request()) // Will suspend until request resolves
  }
```

This code is illegal in TypeScript because `Promise<Value>` is not recognized as `SetStateAction<Value>` subtype. My assumption is that `SetStateAction` should be widened to accommodate for that: `type SetStateAction<Value> = Value | ((prev: Value) => Value) | Promise<Value>`. 

I don't know the codebase well enough to be sure that this is the proper solution.

I can get around this problem creating own `syncAsyncAtom`, but it seems like something that should already be handled at library level.
```
const syncAsyncAtom = <T>(v: T) => atom<T | Promise<T>>(v)
export const dataAtom = syncAsyncAtom<number>(9)
…
    const [data, setData] = useAtom(dataAtom)
    setData(Promise.resolve(45))
…
```

## Link to reproduction
https://codesandbox.io/s/blissful-danny-tlhi2y?file=/src/App.tsx

## Check List

Please do not ask questions in issues.

- [x] I understand this is not a question.

Please include a minimal reproduction.

- [x] I've added a link to a [typescript playground](https://www.typescriptlang.org/play) or [codesandbox](https://codesandbox.io) with a minimal reproduction.

**Comment by dai-shi** [comment, marked as answer]
Yes, you are doing it correctly. It's fully expected.

```ts
const baseAtom = atom<number | Promise<number>>(0)
const Component = () => {
  const [value, setValue] = useAtom(baseAtom)
  const handleClick = () => {
    setValue(Promise.resolve(1))
  }
```

I guess we should have a note for TypeScript users in https://jotai.org/docs/guides/async#async-sometimes. Would you mind opening a PR?

> https://codesandbox.io/s/divine-wildflower-kwrdoh?file=/src/App.tsx:315-364

I wasn't able to find the modified code.

**Reply by fricze** [reply]
Hi @dai-shi, thanks for the reply. I must've linked the wrong sandbox. I updated that now.

I'll open a PR with the documentation change.

Out of curiosity, since it's an expected way of using atoms, don't you think that such solution deserves to have its' own type in the library? Or do you think that typing this properly is easy enough for a developer and so should be left for developer to handle?

**Reply by dai-shi** [reply]
`atom(0)` is inferred as `PrimitiveAtom<number>`. It's wrong for `PrimitiveAtom<number>` to accept `Promise<number>`. It's also wrong for `atom(0)` to be inferred as `PrimitiveAtom<number | Promise<number>>`. Types should be explicit about Promises (In the old versions of jotai, we did something wrong about it.)

So, your `syncAsyncAtom` is a right solution, not just a workaround.

**Reply by fricze** [reply]
Great, I'll add this code to the documentation in incoming PR then.

**Reply by dai-shi** [reply]
> So, your syncAsyncAtom is a right solution, not just a workaround.

But, I would prefer just adding the type annotation, instead of defining `syncAsyncAtom` util which is just overhead code, in docs.

**Reply by fricze** [reply]
Ok
